### PR TITLE
Adds support for prepared query templates.

### PR DESCRIPTION
**Pull request by slackpad**
This is ready for review.

/cc @ryanuber and @sean-

### CODE REVIEW

**Comment by slackpad** [comment]
Here are the benchmarks with the latest version of everything:

```
workpad:prepared_query james$ go test -bench .
PASS
BenchmarkTemplate_CompileSmall-4           10000            105945 ns/op
BenchmarkTemplate_CompileBig-4              5000            271096 ns/op
BenchmarkTemplate_RenderSmall-4            30000             39654 ns/op
BenchmarkTemplate_RenderBig-4              10000            107450 ns/op
ok      github.com/hashicorp/consul/consul/prepared_query       5.191s
```

The small render should be pretty typical of normal users, and 40 us seems pretty reasonable.

File: consul/state/prepared_query.go [hunk 1]
```diff
@@ -16,22 +17,54 @@ func isUUID(str string) bool {
...
+func (s *StateSnapshot) PreparedQueries() (structs.PreparedQueries, error) {
+	queries, err := s.tx.Get("prepared-queries", "id")
 	if err != nil {
 		return nil, err
 	}
-	return iter, nil
+
+	var ret structs.PreparedQueries
+	for wrapped := queries.Next(); wrapped != nil; wrapped = queries.Next() {
+		ret = append(ret, toPreparedQuery(wrapped))
```
Comment (ryanuber) [inline consul/state/prepared_query.go]: Any reason to not just `wrapped.(*queryWrapper).PreparedQuery`? You already handle the nil check before we get here it looks like.

File: consul/state/prepared_query.go [hunk 2]
```diff
@@ -62,14 +95,15 @@ func (s *StateStore) preparedQuerySetTxn(tx *memdb.Txn, idx uint64, query *struc
 	}
 
 	// Check for an existing query.
-	existing, err := tx.First("prepared-queries", "id", query.ID)
+	wrapped, err := tx.First("prepared-queries", "id", query.ID)
 	if err != nil {
 		return fmt.Errorf("failed prepared query lookup: %s", err)
 	}
+	existing := toPreparedQuery(wrapped)
```
Comment (ryanuber) [inline consul/state/prepared_query.go]: This answers my question above about the cast :+1:

File: consul/state/prepared_query.go [hunk 3]
```diff
@@ -16,22 +17,54 @@ func isUUID(str string) bool {
 	return validUUID.MatchString(str)
 }
 
+// queryWrapper is an internal structure that is used to store a query alongside
+// its compiled template, which can be nil.
+type queryWrapper struct {
+	*structs.PreparedQuery
```
Comment (ryanuber) [inline consul/state/prepared_query.go]: Do you ever need to call the methods on a queryWrapper? Trying to understand why the prepared query itself is embedded.

Comment (slackpad) [inline consul/state/prepared_query.go]: Ah true - I think I originally intended to use it as a query, but I could just put the query in a separate member. I can change that if you'd like.

File: consul/prepared_query/template.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	if !ok {
+		return nil, fmt.Errorf("Failed to copy query")
+	}
+
+	// Walk over all the string fields in the Service sub-structure and
+	// parse them as HIL.
+	parse := func(path string, v reflect.Value) error {
+		tree, err := hil.Parse(v.String())
+		if err != nil {
+			return fmt.Errorf("Bad format '%s' in Service%s: %s", v.String(), path, err)
```
Comment (ryanuber) [inline consul/prepared_query/template.go]: Does this render correctly without a space after `Service`?

Comment (slackpad) [inline consul/prepared_query/template.go]: Yeah - this is a little cheesy but all paths have a leading "." so this looks like `Service.Foo`.

File: consul/prepared_query/template.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	// Create a safe match function that can't fail at run time. It will
+	// return an empty string for any invalid input.
+	match := ast.Function{
+		ArgTypes:   []ast.Type{ast.TypeInt},
+		ReturnType: ast.TypeString,
+		Variadic:   false,
+		Callback: func(inputs []interface{}) (interface{}, error) {
+			i, ok := inputs[0].(int)
+			if ok && i < len(matches) {
+				return matches[i], nil
```
Comment (ryanuber) [inline consul/prepared_query/template.go]: Minor and probably an unlikely user error, but it looks like a user could panic Consul here by calling `match -1`.

**Comment by ryanuber** [comment]
Code LGTM! :+1:

**Pull request merged**